frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -531,6 +540,155 @@ describe("QueryBuilder", () => {
     });
   });
 
+  describe("unsaved changes warning", () => {
+    describe("native queries", () => {
+      afterEach(() => {
+        jest.restoreAllMocks();
+      });
+
+      it("shows custom warning modal when leaving edited question via SPA navigation", async () => {
+        const { history } = await setup({
+          card: TEST_NATIVE_CARD,
+          initialRoute: "/home",
+        });
+
+        history.push(`/question/${TEST_NATIVE_CARD.id}`);
+
+        await waitFor(() => {
+          expect(
+            screen.getByTestId("mock-native-query-editor"),
+          ).toBeInTheDocument();
+        });
+
+        const inputArea = within(
+          screen.getByTestId("mock-native-query-editor"),
+        ).getByRole("textbox");
+
+        userEvent.click(inputArea);
+        userEvent.type(inputArea, "0");
+        userEvent.tab();
+
+        history.goBack();
+
+        expect(screen.getByText("Changes were not saved")).toBeInTheDocument();
+        expect(
+          screen.getByText(
+            "Navigating away from here will cause you to lose any changes you have made.",
+          ),
+        ).toBeInTheDocument();
+      });
+
+      it("does not show custom warning modal leaving with no changes via SPA navigation", async () => {
+        const { history } = await setup({
+          card: TEST_NATIVE_CARD,
+          initialRoute: "/home",
+        });
+
+        history.push(`/question/${TEST_NATIVE_CARD.id}`);
+
+        await waitFor(() => {
+          expect(
+            screen.getByTestId("mock-native-query-editor"),
+          ).toBeInTheDocument();
+        });
+
+        history.goBack();
```

Comment:
I really wanted to have [the check suggested by Kelvin in previous PR](https://github.com/metabase/metabase/pull/34192#discussion_r1341034898) here (which would require using `getIsDirty` in `getIsSavedQuestionChanged`) but `Question.prototype.isDirtyComparedTo` returned `true` anyway, see: b1e31e5c02d8174256307584245f377d0591e6f2.

So in case of `QueryBuilder` the `isDirty` flag means "has user interacted with question in any way", while in case of `ActionCreator` the `isDirty` flag means "has action changed in any way".